browser/net/brave_request_handler.cc

```diff
@@ -0,0 +1,298 @@
+  OnReferralHeadersChanged();
+#endif
+}
+
+void BraveRequestHandler::OnReferralHeadersChanged() {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  if (const base::ListValue* referral_headers =
+          g_browser_process->local_state()->GetList(kReferralHeaders)) {
+    base::PostTaskWithTraits(
+        FROM_HERE, {content::BrowserThread::IO},
+        base::Bind(&BraveRequestHandler::SetReferralHeaders,
+                   base::Unretained(this), referral_headers->DeepCopy()));
+  }
+}
+
+void BraveRequestHandler::SetReferralHeaders(
+    base::ListValue* referral_headers) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::IO);
+  referral_headers_list_.reset(referral_headers);
+}
+
+bool BraveRequestHandler::IsRequestIdentifierValid(
+    uint64_t request_identifier) {
+  return ContainsKey(callbacks_, request_identifier);
+  return true;
+}
+
+int BraveRequestHandler::OnBeforeURLRequest(
+    std::shared_ptr<brave::BraveRequestInfo> ctx,
+    net::CompletionOnceCallback callback,
+    GURL* new_url) {
+  if (before_url_request_callbacks_.empty()) {
+    return net::OK;
+  }
+  ctx->new_url = new_url;
+  ctx->event_type = brave::kOnBeforeRequest;
+  callbacks_[ctx->request_identifier] = std::move(callback);
+  RunNextCallback(ctx);
+  return net::ERR_IO_PENDING;
+}
+
+int BraveRequestHandler::OnBeforeStartTransaction(
+    std::shared_ptr<brave::BraveRequestInfo> ctx,
+    net::CompletionOnceCallback callback,
+    net::HttpRequestHeaders* headers) {
+  if (before_start_transaction_callbacks_.empty()) {
+    return net::OK;
+  }
+  ctx->event_type = brave::kOnBeforeStartTransaction;
+  ctx->headers = headers;
+  ctx->referral_headers_list = referral_headers_list_.get();
+  callbacks_[ctx->request_identifier] = std::move(callback);
+  RunNextCallback(ctx);
+  return net::ERR_IO_PENDING;
+}
+
+int BraveRequestHandler::OnHeadersReceived(
+    std::shared_ptr<brave::BraveRequestInfo> ctx,
+    net::CompletionOnceCallback callback,
+    const net::HttpResponseHeaders* original_response_headers,
+    scoped_refptr<net::HttpResponseHeaders>* override_response_headers,
+    GURL* allowed_unsafe_redirect_url) {
+  if (!ctx->tab_origin.is_empty()) {
+    brave::RemoveTrackableSecurityHeadersForThirdParty(
+        ctx->request_url, url::Origin::Create(ctx->tab_origin),
+        original_response_headers, override_response_headers);
+  }
+
+  if (headers_received_callbacks_.empty()) {
+    return net::OK;
+  }
+
+  callbacks_[ctx->request_identifier] = std::move(callback);
+  ctx->event_type = brave::kOnHeadersReceived;
+  ctx->original_response_headers = original_response_headers;
+  ctx->override_response_headers = override_response_headers;
+  ctx->allowed_unsafe_redirect_url = allowed_unsafe_redirect_url;
+
+  base::PostTaskWithTraits(FROM_HERE, {content::BrowserThread::IO},
+                           base::Bind(&BraveRequestHandler::RunNextCallback,
+                                      base::Unretained(this), ctx));
```

Comment:
this `Unretained` looks illegal, thinking about a fix

Reply:
yea, I'm pretty sure you're right. Easy enough to handle with a WeakPtr for BraveRequestHandler

Reply:
Fixed